cupy/_core/dlpack.pyx

```diff
@@ -19,139 +14,213 @@ import warnings
-        shape_strides[n] = array._shape[n]
-    dl_tensor.shape = shape_strides
-    for n in range(ndim):
-        shape_strides[n + ndim] = array._strides[n] // array.dtype.itemsize
 
-    dl_tensor.strides = shape_strides + ndim
-    dl_tensor.byte_offset = 0
+cdef uint8_t get_dlpack_dtype_code(dtype):
+    """Convert NumPy/CuPy to dlpack dtype (kind, without bitsize).
+    """
+    kind = dtype.kind
+
+    if kind == 'u':
+        return <uint8_t>kDLUInt
+    elif kind == 'i':
+        return <uint8_t>kDLInt
+    elif kind == 'f':
+        return <uint8_t>kDLFloat
+    elif kind == 'c':
+        return <uint8_t>kDLComplex
+    elif kind == 'b':
+        return <uint8_t>kDLBool
+    else:
+        raise BufferError('dtype is not supported for dlpack export')
+
 
-    cdef DLDevice* device = &dl_tensor.device
+cdef DLDevice get_dlpack_device(_ndarray_base array):
+    cdef DLDevice device
     cdef bint is_managed
-    cdef int dev_id = array.data.device_id
+
+    device.device_id = array.data.device_id
+
     if not runtime._is_hip_environment:
-        attrs = runtime.pointerGetAttributes(data_ptr)
+        attrs = runtime.pointerGetAttributes(array.data.ptr)
         is_managed = (attrs.type == runtime.memoryTypeManaged)
         if is_managed:
             device.device_type = kDLCUDAManaged
-            dev_id = 0  # make it accessible on CPU too
         else:
             device.device_type = kDLCUDA
     else:
         device.device_type = kDLROCM
-    device.device_id = dev_id
-
-    cdef DLDataType* dtype = &dl_tensor.dtype
-    if array.dtype.kind == 'u':
-        dtype.code = <uint8_t>kDLUInt
-    elif array.dtype.kind == 'i':
-        dtype.code = <uint8_t>kDLInt
-    elif array.dtype.kind == 'f':
-        dtype.code = <uint8_t>kDLFloat
-    elif array.dtype.kind == 'c':
-        dtype.code = <uint8_t>kDLComplex
-    elif array.dtype.kind == 'b':
-        dtype.code = <uint8_t>kDLBool
+
+    return device
+
+
+# The name of this function is following the framework integration guide of
+# TensorComprehensions.
+cpdef object toDlpack(
+    _ndarray_base array, bint use_versioned=True, bint to_cpu=False,
+    bint ensure_copy=False
+) except +:
+    cdef DLManagedTensor* dlm_tensor
+    cdef DLManagedTensorVersioned* dlm_tensor_ver
+    cdef DLTensor* dl_tensor
+    cdef const char *capsule_name
+
+    # Fetch dtype early (as this can raise a BufferError in theory)
+    cdef uint8_t dtype_code = get_dlpack_dtype_code(array.dtype)
+    cdef size_t dtype_itemsize = array.dtype.itemsize
+
+    # Fetch device information since we need it to deal with CPU logic.
+    cdef DLDevice device = get_dlpack_device(array)
+
+    cdef int32_t ndim = array._shape.size()
+
+    if not to_cpu:
+        owner = array
+    elif not ensure_copy and device.device_type == kDLCUDAManaged:
+        # Managed memory is CPU accessible but consumer may expect kDLCPU.
+        owner = array
+        device.device_type = kDLCPU
+        device.device_id = 0
```

Comment:
I worry that after some round-trips the information regarding memory identity would be lost. Do we have to do this? IIRC NumPy also supports `kDLCUDAManaged` since the beginning?

Reply:
Yeah, I wasn't quite sure about what is best here.  The user explicitly asked for "cpu", so I am not super worried about round-tripping.
OTOH, we could clarify that the result must be *compatible* with CPU (i.e. consumer doesn't need to check the device).

Reply:
I am thinking it might be better that we retain this info, as a managed-memory-backed NumPy array could be handled over to another GPU library (other than CuPy) and if the information is retained it'd work nicely back to GPU.